File: Memba.WebControls/MultiUpload.cs
/*
 * Copyright © 2005-2008 Memba SA. All rights reserved.
 * 
 * This file is part of Velodoc XP Edition.
 * 
 * Velodoc XP Edition is open-source software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 2 of the License, or (at your option) any later version.
 * 
 * Velodoc XP Edition is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along with Velodoc XP Edition.
 * If not, see <http://www.gnu.org/licenses/>.
 * 
 * For more information, please contact Memba at <http://www.memba.com>.
 * You can find more information about Velodoc at <http://www.velodoc.com> and <http://www.velodoc.com/help>.
 * You can try Velodoc online at <http://www.velodoc.net>.
 *
*/
using System;
using System.Collections.Generic;
using System.Text;
using System.Web; //AspNetHostingPermission, AspNetHostingPermissionLevel
using System.Web.UI; //Control, IScriptControl, WebResource, Themeable
using System.Web.UI.Design; //UrlEditor
using System.Web.UI.WebControls;
using System.Drawing.Design; //UITypeEditor
using System.ComponentModel; //Category, Description, DefaultProperty
using System.Security.Permissions; //SecurityAction


//See: http://www.asp.net/AJAX/Documentation/Live/tutorials/IScriptControlTutorial1.aspx

[assembly: WebResource("Memba.WebControls.JavaScript.MultiUpload.js", "text/javascript")]

namespace Memba.WebControls
{
    [AspNetHostingPermission(SecurityAction.LinkDemand, Level = AspNetHostingPermissionLevel.Minimal)]
    [AspNetHostingPermission(SecurityAction.InheritanceDemand, Level = AspNetHostingPermissionLevel.Minimal)]
    [DefaultProperty("Text")]
    [Themeable(true)]
    //[Designer("...")]
    //[ParseChildren(true)]
    //[PersistChildren(false)]
    public class MultiUpload : Control, IScriptControl
    {
        #region Private Members
        private string _CssClass;
        private Unit _Height;
        private Unit _Width;
        #endregion

        #region Constructor
        /// <summary>
        /// Constructor
        /// </summary>
        public MultiUpload()
            : base()
        {
        }
        #endregion

        #region Property Accessors
        /// <summary>
        /// Text displayed in the MultiUpload Div
        /// </summary>
        [Category("Appearance")]
        [Description("Text displayed in the MultiUpload Div")]
        [Themeable(false)]
        [Localizable(true)]
        [DefaultValue("")]
        public string Text
        {
            get { return ViewState["Text"] == null ? "" : ViewState["Text"].ToString(); }
            set { ViewState["Text"] = value; }
        }
        /// <summary>
        /// CSS class applied to the MultiUpload Div
        /// </summary>
        [Category("Appearance")]
        [Description("CSS class applied to the MultiUpload Div")]
        [DefaultValue("")]
        public string CssClass
        {
            get { return _CssClass; }
            set { _CssClass = value; }
        }
        /// <summary>
        /// Width of the MultiUpload Div
        /// </summary>
        [Category("Layout")]
        [Description("Width of the MultiUpload Div")]
        [DefaultValue(typeof(Unit), "")]
        public Unit Width
        {
            get { return _Width; }
            set { _Width = value; }
        }
        /// <summary>
        /// Height of the MultiUpload Div
        /// </summary>
        [Category("Layout")]
        [Description("Height of the MultiUpload Div")]
        [DefaultValue(typeof(Unit), "")]
        public Unit Height
        {
            get { return _Height; }
            set { _Height = value; }
        }
        /// <summary>
        /// 
        /// </summary>
        [Browsable(true)]
        public override bool EnableTheming
        {
            get
            {
                return base.EnableTheming;
            }
            set
            {
                base.EnableTheming = value;
            }
        }
        /// <summary>
        /// 
        /// </summary>
        [Browsable(true)]
        public override string SkinID
        {
            get
            {
                return base.SkinID;
            }
            set
            {
                base.SkinID = value;
            }
        }
        #endregion

        #region Control Events
        /// <summary>
        /// 
        /// </summary>
        /// <param name="e"></param>
        protected override void OnPreRender(EventArgs e)
        {
            if (!this.DesignMode)
            {
                // Test for ScriptManager and register if it exists
                ScriptManager objScriptManager = ScriptManager.GetCurrent(Page);

                if (objScriptManager == null)
                    throw new HttpException("A ScriptManager control must exist on the current page.");

                //Register control with ScriptManager
                objScriptManager.RegisterScriptControl(this);
            }

            base.OnPreRender(e); //was on top of the method

        }
        /// <summary>
        /// 
        /// </summary>
        /// <param name="writer"></param>
        protected override void Render(HtmlTextWriter writer)
        {
            writer.AddAttribute(HtmlTextWriterAttribute.Id, this.ClientID);
            writer.AddAttribute(HtmlTextWriterAttribute.Href, "#");
            if (!String.IsNullOrEmpty(_CssClass))
                writer.AddAttribute(HtmlTextWriterAttribute.Class, _CssClass);
            if (!_Width.IsEmpty)
                writer.AddStyleAttribute(HtmlTextWriterStyle.Width, _Width.ToString());
            if (!_Height.IsEmpty)
                writer.AddStyleAttribute(HtmlTextWriterStyle.Height, _Height.ToString());
            //writer.AddStyleAttribute(HtmlTextWriterStyle.Cursor, "pointer");
            writer.RenderBeginTag(HtmlTextWriterTag.A);
            writer.Write(this.Text);
            writer.RenderEndTag(); //A

            writer.AddAttribute(HtmlTextWriterAttribute.Id, this.ClientID + "_Inputs");
            writer.AddStyleAttribute(HtmlTextWriterStyle.Display, "none");
            //but Safari is supposed to prefer the following to display:none
            //writer.AddStyleAttribute(HtmlTextWriterStyle.Position, "absolute");
            //writer.AddStyleAttribute(HtmlTextWriterStyle.Left, "-1000px");
            writer.RenderBeginTag(HtmlTextWriterTag.Div);

            writer.AddAttribute(HtmlTextWriterAttribute.Id, this.ClientID + "_Input_0");
            writer.AddAttribute(HtmlTextWriterAttribute.Name, this.ClientID + "_Input_0");
            writer.AddAttribute(HtmlTextWriterAttribute.Type, "file");
            writer.RenderBeginTag(HtmlTextWriterTag.Input);
            writer.RenderEndTag(); //Input

            writer.RenderEndTag(); //Div

            //Register the Ajax extension properties of the control
            if (!this.DesignMode)
            {
                ScriptManager.GetCurrent(this.Page).RegisterScriptDescriptors(this);
            }

            base.Render(writer); //Was on top of method
        }
        #endregion

        #region IScriptControl interface implementation
        /// <summary>
        /// Protected GetScriptReferences which should be overriden n derived classes
        /// </summary>
        /// <returns></returns>
        protected virtual IEnumerable<ScriptReference> GetScriptReferences()
        {
            yield return new ScriptReference(
                this.Page.ClientScript.GetWebResourceUrl(
                    this.GetType(),
                    "Memba.WebControls.JavaScript.MultiUpload.js"));
        }
        /// <summary>
        /// Protected GetScriptDescriptors which should be overriden n derived classes
        /// </summary>
        /// <returns></returns>
        protected virtual IEnumerable<ScriptDescriptor> GetScriptDescriptors()
        {
            ScriptControlDescriptor objScriptControlDescriptor = new ScriptControlDescriptor("Memba.WebControls.MultiUpload", ClientID);
            objScriptControlDescriptor.AddProperty("text", this.Text);
            objScriptControlDescriptor.AddProperty("cssClass", this.CssClass);

            yield return objScriptControlDescriptor;
        }
        /// <summary>
        /// Public interface for GetScriptReferences
        /// </summary>
        /// <returns></returns>
        IEnumerable<ScriptReference> IScriptControl.GetScriptReferences()
        {
            return GetScriptReferences();
        }
        /// <summary>
        /// Public interface for GetScriptDescriptors
        /// </summary>
        /// <returns></returns>
        IEnumerable<ScriptDescriptor> IScriptControl.GetScriptDescriptors()
        {
            return GetScriptDescriptors();
        }
        #endregion
    }
}

File: Memba.WebControls/JavaScript/MultiUpload.js
/*
 * Copyright © 2005-2008 Memba SA. All rights reserved.
 * 
 * This file is part of Velodoc XP Edition.
 * 
 * Velodoc XP Edition is open-source software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 2 of the License, or (at your option) any later version.
 * 
 * Velodoc XP Edition is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along with Velodoc XP Edition.
 * If not, see <http://www.gnu.org/licenses/>.
 * 
 * For more information, please contact Memba at <http://www.memba.com>.
 * You can find more information about Velodoc at <http://www.velodoc.com> and <http://www.velodoc.com/help>.
 * You can try Velodoc online at <http://www.velodoc.net>.
 *
*/ 
Type.registerNamespace('Memba.WebControls');

//--------------------------------------------
//MultiUploadEventArgs
//--------------------------------------------
Memba.WebControls.MultiUploadEventArgs = function(id, value) {

    Sys.Debug.trace("MultiUploadEventArgs: constructor");

    var e = Function._validateParams(arguments, [
        {name: "id", type: String},
        {name: "value", type: String, mayBeNull: true, optional: true}
    ]);
    if (e) throw e;

    this._id = id;
    this._value = value;
}

Memba.WebControls.MultiUploadEventArgs.prototype = {

    get_id: function() {
        return this._id;
    },

    get_value: function() {
        return this._value;
    }
}

Memba.WebControls.MultiUploadEventArgs.registerClass('Memba.WebControls.MultiUploadEventArgs', null);

//--------------------------------------------
//MultiUpload Control
//--------------------------------------------
Memba.WebControls.MultiUpload = function(element)
{
    Sys.Debug.trace("MultiUpload: constructor");
    
    this._Text = null;
    this._CssClass = null;
    
    this._Anchor = null;
    this._InputsDiv = null;
    this._CurrentInput = null;
    this._CurrentIdNum = null;
    this._Count = null;
    
    Memba.WebControls.MultiUpload.initializeBase(this, [element]);
}

Memba.WebControls.MultiUpload.prototype =
{
    get_text : function()
    {
        return this._Text;
    },

    set_text : function(value)
    {
        if (this._Text != value)
        {
            this._Text = value;
            if (this.get_isInitialized() && (this._Anchor))
                this._Anchor.innerHTML = value;

            // Raise the propertyChanged event
            this.raisePropertyChanged('text');
        }
    },
    
    get_cssClass : function()
    {
        return this._CssClass;
    },
 
    set_cssClass : function(value)
    {
        if (this._CssClass != value)
        {
            this._CssClass = value;
            if (this.get_isInitialized() && (this._Anchor))
                this._Anchor.className = value;

            // Raise the propertyChanged event
            this.raisePropertyChanged('cssClass');
        }
    },
    
    get_count : function()
    {
        return this._Count;
    },

    removeInput : function(id)
    {
        Sys.Debug.trace("MultiUpload: removing input id " + id); 
        var input = $get(id);
        if((input) && (this._InputsDiv) && (this._CurrentInput)
            && (input.id != this._CurrentInput.id) && (input.parentNode.id == this._InputsDiv.id))  
        {
            this._InputsDiv.removeChild(input);
            this._Count--;
        }
        Sys.Debug.trace("MultiUpload: " + this._Count + " files in multi-upload control");
    },
    
    clear : function()
    {
        Sys.Debug.trace("MultiUpload: clear"); 
        while((this._InputsDiv.childNodes) && (this._InputsDiv.childNodes.length > 0))
        {
            this._InputsDiv.removeChild(this._InputsDiv.childNodes[0]);
            this._Count--;
        }
        Sys.Debug.trace("MultiUpload: " + this._Count + " files in multi-upload control");
    },
        
    // browse event
    add_browse: function(handler) {
        /// <summary>Adds a event handler for the browse event.</summary>
        /// <param name="handler" type="Function">The handler to add to the event.</param>
        this.get_events().addHandler("browse", handler);
    },
    remove_browse: function(handler) {
        /// <summary>Removes a event handler for the browse event.</summary>
        /// <param name="handler" type="Function">The handler to remove from the event.</param>
        this.get_events().removeHandler("browse", handler);
    },
    
    initialize : function()
    {
        Sys.Debug.trace("MultiUpload: initialization");
        
        Memba.WebControls.MultiUpload.callBaseMethod(this, 'initialize');
        
        this._Anchor = this.get_element();
        this._Anchor.className = this._CssClass;
        
        this._InputsDiv = $get(this.get_id() + "_Inputs");
        this._CurrentInput = this._InputsDiv.childNodes[0];
        this._CurrentIdNum = 0;
        this._Count = 0;
        
        var onclick = Function.createDelegate(this, function() {
            
            this._CurrentInput.click(); //Does not work in Firefox
            Sys.Debug.trace("MultiUpload: Opening file dialog");
            
            var clickedId = this._CurrentInput.id;
            var clickedValue = this._CurrentInput.value;
            Sys.Debug.trace("MultiUpload: Current input id is " + clickedId);
            Sys.Debug.trace("MultiUpload: Value is " + clickedValue);
            
            //Note: Each time a user clicks the control, this delegate is called twice:
            //First time on clicking the control, before the open file dialog is displayed
            //Second time on clicking Open or Cancel in the open file dialog
            //Note that the first time value is empty, like when clicking cancel
            //Thus teh following test 
            if (clickedValue != '')
            {
                //Add DOM input file element
                this._CurrentIdNum++;
                var newInput;
                //We do not use cloneNode because Netscape copies the value.
                //Neither newInput.name nor newInput.setAttribute("name", ...) work in IE
                //The name should be added in createElement as in the code provided at
                //http://msdn2.microsoft.com/en-us/library/ms534184.aspx
                //See also: http://www.thunderguy.com/semicolon/2005/05/23/setting-the-name-attribute-in-internet-explorer/
                if (Sys.Browser.agent == Sys.Browser.InternetExplorer)
                    newInput = document.createElement('<input name="X"></input>');
                else
                    newInput = document.createElement("input");
                newInput.id = this.get_id() + "_Input_" + this._CurrentIdNum;
                newInput.name = this.get_id() + "_Input_" + this._CurrentIdNum; //Name can be modified though
                newInput.type = "file";
                this._InputsDiv.appendChild(newInput);
                Sys.Debug.trace("MultiUpload: new input added with id: " + newInput.id);
                
                //This should automatically change the scope of the delegate
                //so that a click on the label should now raise a click on the new input
                this._CurrentInput = newInput;
                this._Count++;
                Sys.Debug.trace("MultiUpload: " + this._Count + " files in multi-upload control");

                //Raise the browse event at the end
                var args = new Memba.WebControls.MultiUploadEventArgs(clickedId, clickedValue);
                Sys.Debug.trace("MultiUpload: raising browse event");
                var eventHandler = this.get_events().getHandler("browse");
                if (eventHandler)
                    eventHandler(this, args);
            }
            
            return false; //So that the href="#" in the anchor is not followed
        });
        
        $addHandler(this._Anchor, "click", onclick);
    },
    
    dispose : function()
    {
        Sys.Debug.trace("MultiUpload: disposal");
        
        Memba.WebControls.MultiUpload.callBaseMethod(this, 'dispose');
    }
}

Memba.WebControls.MultiUpload.registerClass('Memba.WebControls.MultiUpload', Sys.UI.Control);

if (typeof (Sys) != 'undefined')
{
    Sys.Application.notifyScriptLoaded();
}


